### ISSUE TITLE
Package scope variable not detected if "use strict" is used

### ISSUE THREAD

**Issue by apendua** [meteor/meteor#5343]
Normally, when I have a package scope "global" variable it would be detected by the builder and added to the package preamble:
```javascript
(function () {

/* Imports */
var Meteor = Package.meteor.Meteor;

/* Package-scope variables */
var MyLocalPackageVariable;

(function () {
  //////////////////////////////////////////////////////
  //
  // packages/custom/main.js
  //
  //////////////////////////////////////////////////////
 // putting 'use strict' here breaks things ...
  MyLocalPackageVariable = {};

  // ...
}());
```
If, however, I add `'use strict'` at the top of the `main.js` file the builder fails to recognize `MyLocalPackageVariable` and it's not added to the preamble anymore unless I explicitly specify it as an exported variable:
```javascript
api.export('MyLocalPackageVariable');
```
which is not an option in this case because I want the variable to remain local to the package.
This results in an error due to strict mode enabled because `MyLocalPackageVariable` is not defined.

There's of course a workaround for this, and that is to use another wrapper:
```javascript
MyLocalPackageVariable = {};
(function () {
  'use strict';
  // ...
}());
```
which is fine but not perfect.

I understand this is happening because for some reason the builder does not recognize global variables which are introduced within blocks with strict mode enabled. I am wondering if this is behavior is intentional. I suppose it's not easy in general to find those variables anyway.

What if we could have an analogue of `api.export` for private variables, e.g.
```
api.private('MyLocalPackageVariable');
```
This would tell the builder to add variable to package preamble (no matter what) but not necessarily export it and make it available outside the package.

**Comment by benjamn** [comment]
Yeah, assigning to undefined (global) variables is a parse-time error in strict mode, which is unfortunate since that's a key part of how the Meteor package system works. I like the idea of adding a `api.private` API, but I'm also hoping that ES2015 modules (planned for Meteor 1.3) make this issue less problematic, since any module inside a package will be able to `import` the `export`ed properties of any other module.

Another solution to this problem might be to transform "global" variable assignments into property assignments on a shared object called something like `packageScope`, so that strict mode would be happy with the resulting code. That's complicated, though, because then all references to those variables would also need to become property access expressions (e.g. `MyLocalPackageVariable` becomes `packageScope.MyLocalPackageVariable` everywhere it appears).

Would you be happy with ES2015 modules, or is it really important to you that the current package system continues to improve?

**Comment by apendua** [comment]
@benjamn Of course having modules will help a lot but it will not solve this particular problem. I the meantime, I realized there's another workaround which feels a lot better than the one I described above. Namely, if I add another file containing "definitions" of all my global variables:
```javascript
// globals.js
MyLocalPackageVariable = null;
```
the `escope` will have a chance to find those implicit globals and so `findAssignedGlobals` will do it's job properly. I think this kind of solution is pretty much equivalent to having `api.private` routine, because it requires you to explicitly define your globals but in a different way.

If we wanted to a have a more generic solution, that would require altering the behavior of `findAssignedGlobals` function. I've done some research based on what you're doing here:

https://github.com/meteor/meteor/blob/5fb0a2316c0ecdc72d483c285708c18ea29cc868/tools/isobuild/js-analyze.js

and I've found out that due to the way `escope` interprets `"use strict"` statements:

https://github.com/estools/escope/blob/master/src/referencer.js#L449

there is no chance global variables are detected unless we manually remove those statements from the `ast` before calling `scopeManager.acquire(ast)`. It seems doable but it also feels like hacking, so I am not 100% sure if it's a good way to go. Ideally we could have `escope` offering an option to ignore `isStrict` property at all, because in our particular use case it's behavior is not the expected one.

**Comment by apendua** [comment]
I also started to think that maybe the way you're dealing with detecting implicit global variables is an overkill. Please note that after all it requires a lot of computation to be done (related to parsing javascript code). Also I think that the package authors are clever enough to be able to properly declare their global variables using some explicit api.

**Comment by stubailo** [comment]
I think this will be resolved in Meteor 1.3, which lets you use ES2015 modules instead, which should work just fine with strict mode.

**Comment by apendua** [comment]
@stubailo Cool, let's see what Meteor 1.3 brings.
